crypto/threads_pthread.c

```diff
@@ -42,12 +42,401 @@
+ */
+void CRYPTO_THREAD_synchronize_rcu(CRYPTO_RCU_LOCK *lock)
+{
+    struct rcu_qp *qp;
+    struct rcu_lock_internal *ilock;
+    uint64_t count;
+    struct rcu_cb_item *cb_items, *tmpcb;
+
+    /*
+     * before we do anything else, lets grab the cb list
+     */
+    ilock = (struct rcu_lock_internal *)lock;
+    cb_items = __atomic_exchange_n(&ilock->cb_items, NULL, __ATOMIC_SEQ_CST);
+
+    qp = update_qp(ilock);
+
+    /*
+     * wait for the reader count to reach zero
+     */
+    do {
+        count = __atomic_load_n(&qp->users, __ATOMIC_SEQ_CST);
+    } while (READER_COUNT(count) != 0);
+
+    /*
+     * retire in order
+     */
+    pthread_mutex_lock(&ilock->prior_lock);
+    while (ilock->next_to_retire != ID_VAL(count))
+        pthread_cond_wait(&ilock->prior_signal, &ilock->prior_lock);
+    ilock->next_to_retire++;
+    pthread_cond_broadcast(&ilock->prior_signal);
+    pthread_mutex_unlock(&ilock->prior_lock);
+
+    retire_qp(ilock, qp);
+
+    /*
+     * handle any callbacks that we have
+     */
+    while (cb_items != NULL) {
+        tmpcb = cb_items;
+        cb_items = cb_items->next;
+        tmpcb->fn(tmpcb->data);
+        CRYPTO_free(tmpcb, NULL, 0);
+    }
+    /*
+     * and we're done
+     */
+    return;
+}
+
+/*
+ * Public WRITE api to schedule a callback after a
+ * completed synchronization
+ */
+void CRYPTO_THREAD_rcu_call(CRYPTO_RCU_LOCK *lock, rcu_cb_fn cb, void *data)
+{
+    struct rcu_cb_item *new =
+        OPENSSL_zalloc(sizeof(struct rcu_cb_item));
+    struct rcu_lock_internal *ilock = (struct rcu_lock_internal *)lock;
+
+    new->data = data;
+    new->fn = cb;
+    new->next = __atomic_exchange_n(&ilock->cb_items, new, __ATOMIC_SEQ_CST);
+}
+
+/*
+ * Public READ/WRITE api to get the value of a pointer
+ */
+void *CRYPTO_THREAD_rcu_uptr_derefrence(uintptr_t *p)
+{
+    return (void *)__atomic_load_n(p, __ATOMIC_ACQUIRE);
+}
+
+/*
+ * Public WRITE api to assign a pointer value
+ */
+void CRYPTO_THREAD_rcu_assign_uptr(uintptr_t *p, uintptr_t *v)
+{
+    __atomic_store(p, v, __ATOMIC_RELEASE);
+}
+
+/*
+ * Public api to allocate a new rcu lock
+ */
+CRYPTO_RCU_LOCK *CRYPTO_THREAD_rcu_lock_new(void)
+{
+    struct rcu_lock_internal *new =
+        OPENSSL_zalloc(sizeof(struct rcu_lock_internal));
+
+    if (new == NULL)
+        return NULL;
+
+    pthread_mutex_init(&new->write_lock, NULL);
+    pthread_mutex_init(&new->prior_lock, NULL);
+    pthread_mutex_init(&new->alloc_lock, NULL);
+    pthread_cond_init(&new->prior_signal, NULL);
+    pthread_cond_init(&new->alloc_signal, NULL);
+    new->qp_group = allocate_new_qp_group(new, 2);
+    if (new->qp_group == NULL) {
+        OPENSSL_free(new);
+        new = NULL;
+    }
+    return (CRYPTO_RCU_LOCK *)new;
+}
+
+/*
+ * Public api to free an rcu lock
+ */
+void CRYPTO_THREAD_rcu_lock_free(CRYPTO_RCU_LOCK *lock)
+{
+    struct rcu_lock_internal *ilock =
+        (struct rcu_lock_internal *)lock;
+
+    /* make sure we're sycchronized */
```

Comment:
synchronized

Reply:
One day I aspire to remember to run ispell on my code prior to committing :)